src/execution/operator/helper/physical_limit_percent.cpp

```diff
@@ -0,0 +1,130 @@
+		if (!op.limit_expression) {
+			this->limit_percent = op.limit_percent;
+			is_limit_percent_delimited = true;
+		} else {
+			this->limit_percent = 100.0;
+		}
+
+		if (!op.offset_expression) {
+			this->offset = op.offset_value;
+			is_offset_delimited = true;
+		} else {
+			this->offset = INVALID_INDEX;
+		}
+	}
+
+	idx_t current_offset;
+	double limit_percent;
+	idx_t offset;
+	ChunkCollection data;
+
+	bool is_limit_percent_delimited = false;
+	bool is_offset_delimited = false;
+};
+
+unique_ptr<GlobalSinkState> PhysicalLimitPercent::GetGlobalSinkState(ClientContext &context) const {
+	return make_unique<LimitPercentGlobalState>(*this);
+}
+
+SinkResultType PhysicalLimitPercent::Sink(ExecutionContext &context, GlobalSinkState &gstate, LocalSinkState &lstate,
+                                          DataChunk &input) const {
+	D_ASSERT(input.size() > 0);
+	auto &state = (LimitPercentGlobalState &)gstate;
+	auto &limit_percent = state.limit_percent;
+	auto &offset = state.offset;
+
+	// get the next chunk from the child
+	if (!state.is_limit_percent_delimited) {
+		Value val = PhysicalLimit::GetDelimiter(input, limit_expression.get());
+		if (!val.is_null) {
+			limit_percent = val.GetValue<double>();
+		}
+		if (limit_percent < 0.0) {
+			throw BinderException("Percentage value(%f) can't be negative", limit_percent);
+		}
+		state.is_limit_percent_delimited = true;
+	}
+	if (!state.is_offset_delimited) {
+		Value val = PhysicalLimit::GetDelimiter(input, offset_expression.get());
+		if (!val.is_null) {
+			offset = val.GetValue<idx_t>();
+		}
+		if (offset > 1ULL << 62ULL) {
+			throw BinderException("Max value %lld for LIMIT/OFFSET is %lld", offset, 1ULL << 62ULL);
+		}
+		state.is_offset_delimited = true;
+	}
+
+	if (!PhysicalLimit::HandleOffset(input, state.current_offset, offset, INVALID_INDEX)) {
+		return SinkResultType::NEED_MORE_INPUT;
+	}
+
+	state.data.Append(input);
+	return SinkResultType::NEED_MORE_INPUT;
+}
+
+//===--------------------------------------------------------------------===//
+// Source
+//===--------------------------------------------------------------------===//
+class LimitPercentOperatorState : public GlobalSourceState {
+public:
+	LimitPercentOperatorState() : chunk_idx(0), limit(INVALID_INDEX) {
+	}
+
+	idx_t chunk_idx;
+	idx_t limit;
+};
+
+unique_ptr<GlobalSourceState> PhysicalLimitPercent::GetGlobalSourceState(ClientContext &context) const {
+	return make_unique<LimitPercentOperatorState>();
+}
+
+void PhysicalLimitPercent::GetData(ExecutionContext &context, DataChunk &chunk, GlobalSourceState &gstate_p,
+                                   LocalSourceState &lstate) const {
+	auto &gstate = (LimitPercentGlobalState &)*sink_state;
+	auto &state = (LimitPercentOperatorState &)gstate_p;
+	auto &limit_percent = gstate.limit_percent;
+	auto &limit = state.limit;
+
+	if (gstate.is_limit_percent_delimited && limit == INVALID_INDEX) {
+		idx_t count = gstate.data.Count();
```

Comment:
Should the count not also take into account the offset?

For example:

```sql
-- returns [0..4]
SELECT * FROM range(10) LIMIT 50%;

-- returns [2..6]
SELECT * FROM range(10) LIMIT 50% OFFSET 2;
```

Reply:
The count should also take into account the offset.

-- returns 5 values, [2..5]
SELECT * FROM range(10) LIMIT 50% OFFSET 2;

I've added this SQL test in test case, e.g.:

query I
SELECT * FROM range(10) LIMIT 50% OFFSET 2
----
2
3
4
5

Reply:
I think original limit implementation is considering as this.

Reply:
That seems a bit odd to me, that means the offset changes the number of result tuples, which is not the case with a regular limit. For example:

```sql
D select * from range(10) limit 5;
┌───────┐
│ range │
├───────┤
│ 0     │
│ 1     │
│ 2     │
│ 3     │
│ 4     │
└───────┘
D select * from range(10) limit 5 offset 5;
┌───────┐
│ range │
├───────┤
│ 5     │
│ 6     │
│ 7     │
│ 8     │
│ 9     │
└───────┘
```

Both return 5 tuples. Now in the current implementation of %, I get the following result:

```sql
D select * from range(10) limit 50%;
┌───────┐
│ range │
├───────┤
│ 0     │
│ 1     │
│ 2     │
│ 3     │
│ 4     │
└───────┘
D select * from range(10) limit 50% offset 5;
┌───────┐
│ range │
├───────┤
│ 5     │
│ 6     │
└───────┘
```

Is this the behavior of the other systems' as well? It seems inconsistent.

Reply:
Okay, I will change the code as your opinion.